core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java

```diff
@@ -0,0 +1,378 @@
+ * holder of Results enumerated here above fully & independently holds complete
+ * ownership of the complete Intellectual Property rights applicable to the whole
+ * of said Results, and may freely exploit it in any manner which does not infringe
+ * the moral rights of the other copyright holders.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.roboconf.dm.templating.internal.templates;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Logger;
+
+import net.roboconf.core.utils.Utils;
+import net.roboconf.dm.templating.internal.TemplatingManager;
+import net.roboconf.dm.templating.internal.helpers.AllHelper;
+import net.roboconf.dm.templating.internal.helpers.IsKeyHelper;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.AbstractFileFilter;
+import org.apache.commons.io.filefilter.CanReadFileFilter;
+import org.apache.commons.io.filefilter.FileFilterUtils;
+import org.apache.commons.io.monitor.FileAlterationListenerAdaptor;
+import org.apache.commons.io.monitor.FileAlterationMonitor;
+import org.apache.commons.io.monitor.FileAlterationObserver;
+
+import com.github.jknack.handlebars.Handlebars;
+import com.github.jknack.handlebars.HandlebarsException;
+import com.github.jknack.handlebars.Template;
+import com.github.jknack.handlebars.io.StringTemplateSource;
+
+/**
+ * A file system watcher dedicated to the Roboconf application templates.
+ * @author Pierre Bourret - Université Joseph Fourier
+ */
+public class TemplateWatcher extends FileAlterationListenerAdaptor {
+
+	private static final ThreadFactory THREAD_FACTORY = new WatcherThreadFactory();
+
+	private final Logger logger = Logger.getLogger( getClass().getName());
+	private final AtomicBoolean alreadyStarted = new AtomicBoolean( false );
+
+	private final ReadWriteLock lock = new ReentrantReadWriteLock( true );
+	private final Handlebars handlebars = new Handlebars();
+
+	// FIXME: should we REALLY keep pre-compiled templates in memory?
+	private final Map<File,TemplateEntry> fileToTemplate = new HashMap<> ();
```

Comment:
We could use `SoftReference<TemplateEntry>` here, and recompile only when the target object has been (garbage) collected. Very basic caching...